components/expression/expression-loader.js

```diff
@@ -52,117 +54,237 @@ genotet.ExpressionLoader.prototype.loadExpressionMatrix_ = function(matrixName,
-    }
-  }else if (identifier.action == 'link') {
-    this.clearLines();
-    this.loadLine(this.lastIdentifier.mat, identifier.source);
-    this.loadLine(this.lastIdentifier.mat, identifier.target);
+/**
+ * Updates the genes in the current expression.
+ * @param {string} method Update method, either 'set' or 'add'.
+ * @param {string} geneRegex Regex that selects the genes to be updated.
+ */
+genotet.ExpressionLoader.prototype.updateGenes = function(method, geneRegex, originParams) {
+  var regex = this.data.matrix.geneRegex;
+  geneRegex = geneRegex.toUpperCase();
+  switch (method) {
+    case 'setGene':
+      // Totally replace the regex.
+      regex = geneRegex;
+      this.data.genes = [];
+      this.data.genes.push(geneRegex);
+      break;
+    case 'addGene':
+      // Concat the two regex's. We need to include the previously existing
+      // genes too so as to find the edges between the new genes and old
+      // ones.
+      regex += '|' + geneRegex;
+      this.data.genes.push(geneRegex);
+      break;
+    case 'removeGene':
+      // Remove the regex.
+      var index = this.data.genes.indexOf(geneRegex);
+      if (index == -1) {
+        genotet.error('invalid gene regex', geneRegex);
+        return;
+      };
+      this.data.genes.splice(index, 1);
+      regex = '';
+      this.data.genes.forEach(function(geneRegexs, i) {
+        regex += geneRegexs + (i == this.data.genes.length - 1 ? '' : '|');
+      }, this);
+      break;
   }
+  var conditionRegex = '';
+  this.data.conditions.forEach(function(conditionNames, i) {
+    conditionRegex += conditionNames + (i == this.data.conditions.length - 1 ? '' : '|');
+  }, this);
+  this.signal('loadStart');
+  var params = {
+    type: 'expression',
+    mat: originParams.matrixName,
+    exprows: regex,
+    expcols: conditionRegex
+  };
+  $.get(genotet.data.serverURL, params, function(data) {
+      // Store the last applied data selectors.
+      _(data).extend({
+        matrixname: originParams.matrixName,
+        geneRegex: regex,
+        conditionRegex: conditionRegex
+      });
+      this.data.matrix = data;
+      this.signal('loadComplete');
+      this.signal('updatePanel');
+    }.bind(this), 'jsonp')
+    .fail(this.fail.bind(this, 'cannot update genes in the expression', params));
 };
 
-LoaderHeatmap.prototype.loadData = function(identifier) {
-  this.parentView.layout.showMsg('Loading...');
-  this.parentView.viewdata = {};
-  this.parentView.viewdata.lineData = new Array();
-  this.lastIdentifier = identifier;
-  if (identifier.name != null && identifier.name != '') {
-    this.loadLine(identifier.mat, identifier.name);
+/**
+ * Updates the conditions in the current expression.
+ * @param {string} method Update method, either 'set' or 'add'.
+ * @param {string} conditionRegex Regex that selects the conditions to be updated.
+ */
+genotet.ExpressionLoader.prototype.updateConditions = function(method, conditionRegex, originParams) {
```

Comment:
I know you used your two functions (of course) but this is not SIGNIFICANT difference. Your update gene/condition routines are extremely similar (aren't they?) and IMO they should be combined into one to prevent duplicated code.